src/core/varinfo.jl

```diff
@@ -55,112 +57,167 @@ type VarInfo
-  if ~overwrite
-    warn("[setval!] you are overwritting values in VarInfo without setting overwrite flag to be true")
-  end
-  vi.vals[getidx(vi, uid)] = val
-end
+getval(vi::VarInfo, view::VarView)       = vi.vals[end][view]
+setval!(vi::VarInfo, val, view::VarView) = vi.vals[end][view] = val
 
-getsym(vi::VarInfo, vn::VarName) = vi.uids[getidx(vi, vn)][2]
-getsym(vi::VarInfo, uid::Tuple) = vi.uids[getidx(vi, uid)][2]
+getsym(vi::VarInfo, vn::VarName) = vi.vns[getidx(vi, vn)].sym
 
 getdist(vi::VarInfo, vn::VarName) = vi.dists[getidx(vi, vn)]
-getdist(vi::VarInfo, uid::Tuple) = vi.dists[getidx(vi, uid)]
-setdist!(vi::VarInfo, dist, vn::VarName) = vi.dists[getidx(vi, vn)] = dist
-setdist!(vi::VarInfo, dist, uid::Tuple) = vi.dists[getidx(vi, uid)] = dist
 
 getgid(vi::VarInfo, vn::VarName) = vi.gids[getidx(vi, vn)]
-getgid(vi::VarInfo, uid::Tuple) = vi.gids[getidx(vi, uid)]
+
 setgid!(vi::VarInfo, gid, vn::VarName) = vi.gids[getidx(vi, vn)] = gid
-setgid!(vi::VarInfo, gid, uid::Tuple) = vi.gids[getidx(vi, uid)] = gid
 
 istransformed(vi::VarInfo, vn::VarName) = vi.trans[getidx(vi, vn)]
-istransformed(vi::VarInfo, uid::Tuple) = vi.trans[getidx(vi, uid)]
-settrans!(vi::VarInfo, trans, vn::VarName) = vi.trans[getidx(vi, vn)] = trans
-settrans!(vi::VarInfo, trans, uid::Tuple) = vi.trans[getidx(vi, uid)] = trans
+
+# X -> R for all variables associated with given sampler
+function link(_vi, spl)
+  vi = deepcopy(_vi)
+  gkeys = groupvns(vi, spl)
+  for k in gkeys
+    dist = getdist(vi, k)
+    vi[k] = link(dist, reconstruct(dist, getval(vi, k)))
+    vi.trans[getidx(vi, k)] = true
+  end
+  vi
+end
+
+# R -> X for all variables associated with given sampler
+function invlink(_vi, spl)
+  vi = deepcopy(_vi)
+  gkeys = groupvns(vi, spl)
+  for k in gkeys
+    dist = getdist(vi, k)
+    vi[k] = invlink(dist, reconstruct(dist, getval(vi, k)))
+    vi.trans[getidx(vi, k)] = false
+  end
+  vi
+end
+
+vns(vi::VarInfo) = Set(keys(vi.idcs))            # get all vns
+syms(vi::VarInfo) = map(vn -> vn.sym, vns(vi))  # get all symbols
 
 # The default getindex & setindex!() for get & set values
-Base.getindex(vi::VarInfo, vn::VarName) = getval(vi, vn)
-Base.getindex(vi::VarInfo, uid::Tuple) = getval(vi, uid)
-Base.setindex!(vi::VarInfo, val, vn::VarName) = setval!(vi, val, vn, true)
-Base.setindex!(vi::VarInfo, val, uid::Tuple) = setval!(vi, val, uid, true)
+# NOTE: vi[vn] will always transform the variable to its original space and Julia type
+Base.getindex(vi::VarInfo, vn::VarName) = begin
+  @assert haskey(vi, vn) "[Turing] attempted to replay unexisting variables in VarInfo"
+  dist = getdist(vi, vn)
+  r = reconstruct(dist, getval(vi, vn))
+  r = istransformed(vi, vn) ? invlink(dist, r) : r
+end
 
-uids(vi::VarInfo) = Set(keys(vi.idcs))            # get all uids
-syms(vi::VarInfo) = map(uid -> uid[2], uids(vi))  # get all symbols
+Base.setindex!(vi::VarInfo, r, vn::VarName) = begin
+  dist = getdist(vi, vn)
+  setval!(vi, vectorize(dist, r), vn)
```

Comment:
Do you mean by transforming `r` to the corresponding format indicated in `VarInfo`? Actually this something not clear to me because we don't really know whether `r` is transformed or not. Shall we assume `r` is in the original space then?